### ISSUE TITLE
Remaining work to get pull diagnostics + LSP fully enabled, and all push-diagnostics + error list code removed.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#62772]
1. Need to figure out how someone can configure suppressions from the error list if we don't own the error list.
2. Need to remove IDiagnosticService.DiagnosticsUpdated.  Currently this is used by tagger (but is not needed if tagger is just pulling), and by typescript.  David believes TS just uses this for LSP-push-diags in liveshare and can likely be removed since we have worked with them to support pull diagnostics.
3. Move IDiagnosticService.GetDiagnosticsAsync to be snapshot based.  Currently it takes in Workspace+ProjId/DocId.  This is non-sensical as that means you may get random results depending on the state of Workspace.CurrentSnapshot.  This needs to move to take Project/Document instead.
4. Need LSP pull diagnostics to be category-based, so we can report syntax before semantics.

See also: https://github.com/dotnet/roslyn/pull/62741, https://github.com/dotnet/roslyn/pull/62735 and https://github.com/dotnet/roslyn/pull/62771

**Comment by CyrusNajmabadi** [comment]
@tmat @mavasani For visibility.  We're very close on this :)

**Comment by tmat** [comment]
Cool. My experience with pull diagnostics in Roslyn.sln is not great though. They seem to be lagging quite a bit. That needs to be addressed imo, before we switch.

**Comment by tmat** [comment]
I haven't had chance to do much work in Roslyn.sln since.

**Comment by dibarbet** [comment]
thanks @mavasani - I think that is an optimization I'll need to port over.

**Comment by dibarbet** [comment]
@mavasani hmm - in the incremental analyzer implementation, it looks like it ignores the method node we find during the differencing - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs,28

**Comment by mavasani** [comment]
> @mavasani hmm - in the incremental analyzer implementation, it looks like it ignores the method node we find during the differencing - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs,28

Wow, that seems like a big regression. I need to look at the history to figure out when/why this was removed.

**Comment by mavasani** [comment]
> in the incremental analyzer implementation, it looks like it ignores the method node we find during the differencing - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs,28

Tagging @heejaechang for historical context.

I investigated this a bit more and it seems the IDE DiagnosticIncrementalAnalyzer's v1 implementation performed method body based analysis, but this was never implemented for the v2 engine. Below are the implementations from the same repo snapshot:

https://github.com/dotnet/roslyn/blob/82c9de92110a90088e3e03d5c86b072a795ff4e8/src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs#L278-L286

https://github.com/dotnet/roslyn/blob/82c9de92110a90088e3e03d5c86b072a795ff4e8/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs#L16-L28

v2 engine was written and replaced the v1 engine about 6 years ago, so this is not a recent behavior change. @heejaechang do you recall if this method body edit optimization was removed due to bugs/flakiness or was it just pending work that never got done? If it is the latter, then it might make sense to only port this optimization to the LSP handler. The good part here is:
1. This would open up scope for quite a decent performance gain when moving to LSP pull diagnostics for method body editing scenarios.
2. This optimization would be an enhancement, rather than a parity requirement for enabling pull diagnostics by default.

I do notice that for the light-bulb path we do use `IBuiltInAnalyzer.GetAnalyzerCategory()` to only execute span based analyzers for the line span, which is good:

https://github.com/dotnet/roslyn/blob/3fe68247686c879325eabdb3eddc4f11acc8ecb7/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs#L173-L192

**Comment by CyrusNajmabadi** [comment]
@mavasani in a pull-diags world, if there is any sort of caching, i would assume that adding this optimization back in would be fairly trivial.  we should know the Doc/Tree the last pulled diags were against.  the next pull can attempt to diff the current tree against the previous one.  If teh diff is entirely within a method, we can keep all the diags outside of that, and only recompute what is inside right?

Basically, my expectation is that pull ideally makes this easier to add as an invisible optimization (outside of perf speedup).

**Comment by heejaechang** [comment]
https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs,263

this should call the body only analysis for open file?

**Comment by sharwell** [comment]
If we did end up shipping for 6 years without the method body detection enabled, I would encourage we just make that state permanent, since it allows us to finally remove `IBuiltInAnalyzer`.

**Comment by CyrusNajmabadi** [comment]
note that in that time we have had an enormous amount of complaints about performance :)

**Comment by mavasani** [comment]
>https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs,263
this should call the body only analysis for open file?

@heejaechang It does, and solution crawler is fine here. The issue is the DiagnosticIncrementalAnalyzer ignore the body node: https://github.com/dotnet/roslyn/blob/843f8db03084db9c6e8c1abea5681caf1d535d23/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs#L28-L29

> If we did end up shipping for 6 years without the method body edit detection enabled, I would encourage we just make that state permanent, since it allows us to finally remove IBuiltInAnalyzer.

@sharwell Note that for light-bulb path we do use `IBuiltInAnalyzer.GetAnalyzerCategory()` to only execute span based analyzers for the line span:

https://github.com/dotnet/roslyn/blob/3fe68247686c879325eabdb3eddc4f11acc8ecb7/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs#L173-L192

If we remove that logic, we will start executing SimplifyTypeName analyzer on the the entire file every time Ctrl + Dot is invoked. We certainly don't want that.

**Comment by CyrusNajmabadi** [comment]
I also wouldn't say that we haven't noticed.  Perf has been a perpetual problem.  It's just never been clear what is at fault.  I know that i had to add back smarts to scope thigns back to method bodies to help speed up perf in completion lists for example :)  i think we're just not in a good place perf-wise, with lots of components being problematic, so it's easy for regressions to happen that just get added to the pile of problems :)

**Comment by heejaechang** [comment]
interesting... what I remember is in proc diagnostic analyzer uses method body (since in proc handles open file case) and out of proc doesn't care about method body (since it only handles closed file and compilation diagnostics). and I believe out of proc doesn't even have concept of open or visible files ..

but certainly, a bug could have introduced..

**Comment by CyrusNajmabadi** [comment]
Closing as this has all been done.
